### ISSUE TITLE
Allow setting the Activity properties after Activity creation and before starting it when using ActivitySource

### ISSUE THREAD

**Issue by macrogreg** [dotnet/runtime#42784]
*The description and the Proposal edited by @tarekgh. The original description is moved down in the comment https://github.com/dotnet/runtime/issues/42784#issuecomment-777143962*

ActivitySource provides the method StartActivity which will try to create and start the Activity object. There is a scenario that needs to add some more properties to the Activity object after it is created and before it gets started. 
The reason is the caller wants to avoid creating the objects need to set to the Activity before ensuring the Activity object is created. It is important to have these properties set before starting the Activity so the listeners of the Start event will get the full set of the data.

# Proposal 

```C#
namespace System.Diagnostics
{
    public sealed class ActivitySource : IDisposable
    {
        // Alternative proposed name is CreateUnstartedActivity
        public Activity? CreateActivity(
                             string name,
                             ActivityKind kind, 
                             ActivityContext parentContext,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null);

        public Activity? CreateActivity(
                             string name, 
                             ActivityKind kind,
                             string parentContext,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null);
    }
}

```

# Example

```C#
    using (Activity activity = activitySource.CreateActivity("SomeName", ActivityKind.Server, context))
    {
        activity?.AddBaggage("db.type", "mongo");
        activity?.Start();

        // Do any necessary work      
    }
```

**Comment by macrogreg** [comment]
Some tags to raise awareness for folks involved into relevant work streams:
@cijothomas  @noahfalk  

Thanks! :)

**Comment by tarekgh** [comment]
I don't think it is intended to support all possible scenario in ActivitySource in first version. Your scenario can be considered in the future releases. You still have the option to use old way of creating the Activity and using DiagnosticSource. This still work.

For tags allocation, you can still can avoid the allocation before creating the activity by setting `ActivityCreationOptions.SamplingTags` in the listener callbacks. This will still allocate at that time but for sure it will be determined the Activity is going to be allocated anyway.

Is too late to change any APIs in the 5.0 release.

**Comment by macrogreg** [comment]
>  You still have the option to use old way of creating the `Activity `and using `DiagnosticSource`. This still work.

From the perspective of a telemetry producer - yes. But a producer does not know whether this scenario is important for any actual consumer. And if a producer who followed the existing guidance wants to transition to the "new way" while enabling all previous scenarios, they cannot. I completely realize the current spot in the release cycle, but if this item was to be tracked for prioritization in the next cycles, then, for the record, I think that the workaround you mention does not apply. :)

> For tags allocation, you can still avoid the allocation before creating the activity by setting `ActivityCreationOptions.SamplingTags` in the listener callbacks. 

Thanks for pointing it out. I had missed the existence of that property. However, having had a look at it, I do not fully understand its use. Is there a doc or an example? What I mean above is that when invoking `ActivitySource.StartActivity(..)` you need to pass an `IEnumerable` object on the heap. So the allocation will occur when creating that object while the sampling decision is taken within the scope of the `ActivitySource.StartActivity(..)` method. So regardless of whether or not you decided to allocate the activity, you will have to allocate the tags collection object.

Please let me know if you guys consider these Activity-related submissions as a valuable input for forward-looking planning / for the next cycle. I do not want to create a distraction if it is not productive. :)

Thanks!

**Comment by macrogreg** [comment]
> Agree. What i was trying to say is, this is not supported scenario yet with ActivitySource. I was pointing if someone blocked for now can consider use the old way.

Gotcha, thanks!

> For ActivityCreationOptions.SamplingTags, the idea is when you are calling ActivitySource.StartActivity you'll not pass any tags at that time but in the listener you'll have a chance to add the tags during then.

Ah, understood. This is a different scenario. I was talking about saving the allocation that is only required only to pass parameters to StartActivity(..) 

> If the allocation is very critical to your scenario, you may consider having IEnumerable thread local object (e.g. List<>) which you can use for all subsequent calls on same thread. 

That might work, but the complexity of implementing such cache and the perf overhead required for cleaning it before each use make this feel concerning. Of course, .NET5 is done. I would love a performance-focused overload similar to what I mentioned above to be considered for the next release.

Thanks!

**Comment by cijothomas** [comment]
> StartActivity(..) overloads that have parameters like string tag1Key, string tag1Value, ... for the first few tags.

This sounds like a good improvement. In most cases the number of initial tags/attributes is very low (or none). More tags are added after actually creating the activity and knowing the sampling decision.
``` 
(if activity != null && activity.IsAllDataRequested == true) 
{
  add tags.
}
```

**Comment by noahfalk** [comment]
Thanks @macrogreg!

> System.Diagnostics.ActivitySource.StartActivity(..) is missing an overload that takes Baggage items

StartActivity() was intended to take the minimum amount of data necessary for sampling. Adding more types of data trades off performance for more flexibility in sampling and richer data available on the start callback. Another direction is to create an API that returns an unstarted Activity allowing Start() to be decoupled from sampling. That option is trading off more complicated API usage for richer data available on the start callback. I'm not opposed to this in principle, but they first hurdle to clear would be to understand why do we need specific data attached to a Start callback if that same data will be available later in the Stop callback? 

> However, this will require that an enumerable is allocated before StartActivity(..) is called

We added the overload which took tags because OpenTelemetry defined that it must be possible to give tags to the sampling algorithm. My hope is that in practice it is a rare case that the sampling algorithm needs to consider tags because doing so implies you are paying perf costs to generate them and analyze them even for sampled-out activities. The higher performing route would be to determine in advance that the tags aren't relevant for sampling and provide them only after an Activity has been sampled in using Activity.SetTag(). If there is a need for higher performance scenarios where certain tags are available to the sampler or available in the Start() callback I'm not opposed to optimizations you suggested, but I'd like to be able to tie to some specific customer or use-case that the work is improving.

**Comment by macrogreg** [comment]
> I'm not opposed to this in principle, but they first hurdle to clear would be to understand why do we need specific data attached to a Start callback if that same data will be available later in the Stop callback?

Agree. At this point I am merely pointing out that the [.NET guidance pre .NET 5](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#starting-and-stopping-activity) was to include baggage for the Activity Start callback and the new API doe snot allow it. We may decide that the guidance was wrong or that the different ins not important and does not justify more complexity. Both is OK. Just feels like it should be an explicit decision, rather than an oversight. :)

> We added the overload which took tags because OpenTelemetry defined that it must be possible to give tags to the sampling algorithm. My hope is that in practice it is a rare case that the sampling algorithm needs to consider tags because doing so implies you are paying perf costs to generate them and analyze them even for sampled-out activities. The higher performing route would be to determine in advance that the tags aren't relevant for sampling and provide them only after an Activity has been sampled in using Activity.SetTag(). If there is a need for higher performance scenarios where certain tags are available to the sampler or available in the Start() callback I'm not opposed to optimizations you suggested, but I'd like to be able to tie to some specific customer or use-case that the work is improving.

Ok, that makes sense. However, in the context of the aforementioned old-API-guidance, it was not clear to me that the purpose of the overload that takes tags was to make tags available to the sampler, not the Start event handler.

I think others may also be confuted in a similar manner. Perhaps we can add guidance to the docs to say something along the lines that the recommended pattern is to:

* Only include info into the StartActivity(..) that is relevant for sampling
* Start event handlers should _not_ expect that the information that is logically available at the beginning of the operation represented by an Activity is actually attached to the Activity when the Start event is raised. (E.g., input parameters).
* Both should be attached to the Activity when the End event is raised:
   * tracing-relevant information that is readily available _at the beginning_ of the sub-operation represented by an Activity (e.g. input parameters)
   * as well as tracing-relevant information that is only available _at the end_ of the sub-operation represented by an Activity (e.g. results) 
* This is different from previously published DiagnosticSource examples, where the input parameters relevant information was passed to the start event

That would clear things up. :)
Note, however, this would prevent scenarios where an exporter wishes to record as useful events as possible both, at the start and the end of an Activity.

**Comment by noahfalk** [comment]
> That would clear things up. :)

Agreed. This matches guidance I had in mind but we hadn't explicitly documented it as far as I know. I created https://github.com/dotnet/docs/issues/20825 to track doing that.

> [.NET guidance pre .NET 5](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#starting-and-stopping-activity)

Yeah I consider that guidance out-of-date now. Perhaps we should use this issue to track deprecating that info and redirecting it to the official docs once we've got that complete?

**Comment by macrogreg** [comment]
This works from my perspective. To inform this whole thing, here is a curious piece of historical trivia.

I was not involved into the creation of Activity class, but if I understand correctly (I would be curious to learn for sure), it was inspired by a an internal logging library that were used by some Application Insights backed services. Or perhaps it was not, but the APIs look similar and there is some overlap in the people involved, so who knows. :)

That other library already operated with the term Activity, but the focus was more around structured logging, rather than distributed context. You started and finished Activities to describe sub-operations of your business logic. You could also attach properties (=tags) to the Activities. The system flew the Current activity across async boundaries, in 4.5 - using a custom synchronization context; later - using AsyncLocal. A guy named Bogdan and another guy called Greg created that system in 2013/14. Not sure if you know Bogdan. Greg is pleased to have this conversation with you. :)

The activities were written to a configurable sink. Back then is was a file or an early re-incarnation of MDS. Later, it was Application Insights. We had to decide, at what point would we record a persistent event:
- when as activity is completed?
- every time a property is changed?
.. this was changed many times and was made configurable, but in the end most services adopted a model where a record was emitted at the start and at the end of each activity, where the information available at the start (e.g. input parameters) was included into the start record. The reason for that was that it was very useful to do log-based debugging in case of service issues: you could trace the execution flow by reading activity events and if a problem occurred during an activity, you still had a record of it starting with all the relevant info.

The reason I am dwelling in this is because I believe that over time, the systems we build may evolve so that these scenarios may move into focus again.

.NET Activities are not used for structured logging today (I mean structured as in business-logic structure of the app, not structure of a log entry). Instead, they are focused on understanding the distributed flow and on propagating distributed tracing context. For that - indeed - I cannot think of why it would be necessary to persist an activity started event. However, as I envision the future, I cannot help the feeling that logging and tracing want to be One thing. 
Could it be that in a few years we will arrive at a spot where the ILogger just adds events (tags, properties) onto the current activity? Or perhaps, in a reversed scenario, a tracing backend will be the same as a logging backend with ActivityStartEvent, InformationalLogEvent, ActivityEndEvent, ErrorLogEvent, etc., just being different values in the Kind column of an all-encompassing application log stream?

So, yes, thinking of the scenarios of this year's release, there is no need to any information to the activity before it starts, beyond what is required for sampling. The API for .NET 5 is done and its a good one.

But going forward, allowing the exporter to log full-fidelity activity-start events, and encouraging telemetry producers to emit telemetry in a way that makes it possible may lead to a more flexible and powerful telemetry platform. Just my 5 cents. :)

**Comment by noahfalk** [comment]
> Not sure if you know Bogdan

Bogdan Berce perhaps? I've worked with him : )

> Greg is pleased to have this conversation with you

I'm liking the literary flair!

> I would be curious to learn for sure

My understanding was that Vance Morrison was responsible for the original Activity API (quite possibly based on your library, I wasn't there to say?) perhaps in conjunction with @brianrob. I started getting involved with it more heavily during early .Net Core 3.0 as we added the W3C support and then took over when Vance retired.

> But going forward, allowing the exporter to log full-fidelity activity-start events, and encouraging telemetry producers to emit telemetry in a way that makes it possible may lead to a more flexible and powerful telemetry platform

I'm happy to see how usage and requests evolve. If users want start events that carry precise user-defined payloads of additional information I imagine we can continue to massage the API to support that. Activity (and distributed tracing in general) I think of as a messy, pragmatic and useful diagnostic tool that has been pushed to do a little of everything. I just hope we can keep stewarding it well and keep sight of the tradeoffs as we seem to evolve it towards new goalposts with every release.

> The API for .NET 5 is done and its a good one.

Thanks! That is nice to hear : )

**Comment by macrogreg** [comment]
Yes, that Bogdan. :)

 > My understanding was that Vance Morrison was responsible for the original Activity API (quite possibly based on your library, I wasn't there to say?) perhaps in conjunction with @brianrob. I started getting involved with it more heavily during early .Net Core 3.0 as we added the W3C support and then took over when Vance retired.

Oh, I see. Thanks for the context. I was thinking that Vance was mainly working on DiagnosticSource and @lmolkova and @SergeyKanzhelev added Activity [some time later](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/4ab7cd3837ae5039128f26fd3aa9ab8c0c6d9861/src/OpenTelemetry.DynamicActivityBinding/OpenTelemetry.DynamicActivityBinding/internal/DynamicLoader.cs#L26). (Not that it matters, I just like history :) .)

> I'm happy to see how usage and requests evolve. If users want start events that carry precise user-defined payloads of additional information I imagine we can continue to massage the API to support that. Activity (and distributed tracing in general) I think of as a messy, pragmatic and useful diagnostic tool that has been pushed to do a little of everything. I just hope we can keep stewarding it well and keep sight of the tradeoffs as we seem to evolve it towards new goalposts with every release.

Yes. I think we are now in agreement on .NET 5's guidance, I just wanted to make it explicit. 
Right now we are in into-the-future-looking land and my goal is to raise some early suggestions and directions for everybody's consideration.

**Comment by jbogard** [comment]
I just hit this today - updating from `DiagnosticListener` to `ActivitySource` and realizing none of my tags made it to the start events, and then out to the samplers/exporters. Not a show-stopper but a bit different in current behavior.

**Comment by tarekgh** [comment]
@jbogard could you please clarify more regarding `none of my tags made it to the start events`? did you pass the tags to `ActivitySource.StartActivity`?

**Comment by jbogard** [comment]
No, because it wasn't exactly clear whether I was supposed to pass tags in that overload, or check to see if `activity.IsAllDataRequested` and add the tags there. It looks like the OTel repo by and large adds tags *only* after the sampling decision is made. That means my tags only show up on Stop.

**Comment by tarekgh** [comment]
CC @cijothomas to comment on OTel.

**Comment by cijothomas** [comment]
The title of this issue says it is about Baggage - is the question about Baggage? The Activity.Baggage != Otel Baggage specification. To use Baggage, one must use Otel API.

**Comment by jbogard** [comment]
@cijothomas this was in reply to my comment above, about when to populate an `Activity` with tags. At the start of the activity, or after it is created. Baggage was just something else that is not available in the start event, only at Stop.

**Comment by cijothomas** [comment]
Oh sorry! I am not sure what is the question for OpenTelemetry - could you rephrase the question please?

**Comment by jbogard** [comment]
It's not a question for OpenTelemetry. It's a comment about the API. Here's my MongoDB driver instrumentation when using `ActivitySource`:

```csharp
var activity = ActivitySource.StartActivity(ActivityName, ActivityKind.Client);

if (activity == null)
{
    return;
}

activity.DisplayName = $"mongodb.{@event.CommandName}";

if (activity.IsAllDataRequested)
{
    activity.AddTag("db.type", "mongo");
    activity.AddTag("db.instance", @event.DatabaseNamespace.DatabaseName);
    // many more tags
```

I add all the tags *after* the activity is started. Which means they won't be available to an `ActivityListener`, and downstream tracing clients.

Previously, the "start" event via `DiagnosticListener` would carry a payload of context that I would then use to enrich the `Activity` further before it got recorded by a telemetry client (OpenTelemetry in my case).

```csharp
private void Handle(CommandStartedEvent @event)
{
    var activity = new Activity(ActivityName);

    _diagnosticListener.OnActivityExport(activity, @event);

    if (_diagnosticListener.IsEnabled(ActivityStartEventName, @event))
    {
        _diagnosticListener.StartActivity(activity, @event);
    }
    else
    {
        activity.Start();
    }
```

Then, my diagnostic subscriber, I would add the tags if all data was requested, then bridge in this case to OTel (though that's immaterial):

```csharp
public override void OnStartActivity(Activity activity, object payload)
{
    if (!(payload is CommandStartedEvent message))
    {
        return;
    }

    activity.DisplayName = $"mongodb.{message.CommandName}";

    if (activity.IsAllDataRequested)
    {
        activity.AddTag("db.type", "mongo");
        activity.AddTag("db.instance", message.DatabaseNamespace.DatabaseName);
        // many more tags
    }
    _activitySource.Start(activity, ActivityKind.Client);
```

I don't record the activity to a telemetry client until after I've added all the tags based on the context being passed through. This is what I meant by "tags are missing". The events are recorded in the `ActivitySource` method, but because I only add all the tags if `IsAllDataRequested`, I've lost the opportunity to have these tags added or context available in the "Start" event.

**Comment by macrogreg** [comment]
Thanks, @jbogard ,  yes, you describe the same high-level challenge. :)
I have a side question:

> Here's my MongoDB driver instrumentation when using ActivitySource

Is this the official MongoDB client, meaning MongoDB will get support of ActivitySource-based racing, or is this your own project?

Thanks!

**Comment by jbogard** [comment]
It's my own project, which added `DiagnosticListener`, which I'm now updating for `ActivitySource` [MongoDB.Driver.Core.Extensions.DiagnosticSources](https://github.com/jbogard/MongoDB.Driver.Core.Extensions.DiagnosticSources). Though I do have an open JIRA ticket for `System.Diagnostics`.

**Comment by tarekgh** [comment]
@jbogard when you are doing `var activity = new Activity(ActivityName);` this will make `activity.IsAllDataRequested` is always true by default. Also `IsAllDataRequested` is a new property and never existed before. So how this used to work before?

**Comment by jbogard** [comment]
Ah, right I was referencing the 5.0 RC. The 4.7 version would simply add all the tags to the activity.

**Comment by tarekgh** [comment]
@jbogard but still your logic in RC was adding all tags to the activity too.

I think the ask here now is a feature request to have a way adding tags/baggage or anything else conditionally to the Activity after it is created and before it is started. I think this is reasonable ask to track.

**Comment by jbogard** [comment]
Yes it was adding the tags in a `DiagnosticListener` of the activity events. That piece of code leveraged OpenTelemetry which encapsulated a lot of the observable business. So it was *after* the diagnostic event was raised.

**Comment by noahfalk** [comment]
Hi @jbogard, sorry this issue has been quiet a bit. @tarekgh and I were taking a look at issues like this one that came in late in .NET 5 to see what we could do in .NET 6. I could see a few potential paths forward here:
1. We change the past guidance to indicate that tags/baggage/other Activity content is not fully populated in the Start event. Instrumentors would add it after the Activity is started and Listeners would pick it up during the Stop event. For people who were using DiagnosticListener in the past to transmit Activities that are now using ActivitySource this is certainly a change in behavior but perhaps not an impactful one?
2. .NET could create a new API such as ActivitySource.CreateUnstartedActivity(...) that has the same parameters as StartActivity(). Then you could write code that looked like this which retains behavior closer to what you had before:

````C#
// this invokes ActivityListener.Sample callback but not ActivityListener.Start
var activity = ActivitySource.CreateUnstartedActivity(ActivityName, ActivityKind.Client);

if (activity == null)
{
    return;
}

activity.DisplayName = $"mongodb.{@event.CommandName}";

if (activity.IsAllDataRequested)
{
    activity.AddTag("db.type", "mongo");
    activity.AddTag("db.instance", @event.DatabaseNamespace.DatabaseName);
    // many more tags
}

// now the ActivityListener.StartActivity callback gets invoked and the full set of Tags, Baggage, etc 
// that you want to provide can be present
activity.Start();
````
3. Some other alternative that you have in mind?

I think the discussion before the holidays was converging towards option (2), but I wanted to confirm. If so we can firm up the details and get the ball rolling on this API. Thanks!

**Comment by noahfalk** [comment]
Slightly off topic, but about a week or two ago I was following random tweets which lead me to https://jimmybogard.com/building-end-to-end-diagnostics-activitysource-and-open/ and https://jimmybogard.com/activitysource-and-listener-in-net-5/. It was beautifully written. Thank you very much for putting it out there!

**Comment by tarekgh** [comment]
@jbogard could you please let's know if you have any comment before we proceed with the proposal.

**Comment by jbogard** [comment]
I'd lean towards option 2 as well - adding more parameters to `ActivitySource.StartActivity` would be less than ideal.

**Comment by cijothomas** [comment]
Consider `ActivitySource.CreateActivity()`, along with existing `ActivitySource.StartActivity()` as alternate name to `ActivitySource.CreateUnstartedActivity()`

**Comment by tarekgh** [comment]
I have added the proposal in the top. We should discuss the naming in the design meeting. Thanks allot for your feedback.

**Comment by cijothomas** [comment]
minor: The title of this issue can be confusing as the new API is about slightly different things. Could you create a new issue with the proposal, and link it from there for easy read by anyone new. :)

**Comment by tarekgh** [comment]
> minor: The title of this issue can be confusing as the new API is about slightly different things. Could you create a new issue with the proposal, and link it from there for easy read by anyone new. :)

We can rename the title and possibly the description if @macrogreg not mind. I am not sure creating a new issue and linked to this one would help much as all discussions are listed here.

**Comment by tarekgh** [comment]
Moving the original description added by @macrogreg here for the record as we'll need to modify the description up to match what have discussed and decided on. 

*@macrogreg original description*

I think there is a scenario supported by the pre-.NET-5 Activity APIs based on `DiagnosticSource` that is not supported in the new `ActivitySource` based APIs:

**`System.Diagnostics.ActivitySource.StartActivity(..)` is missing an overload that takes Baggage items.**

Consider .NET's own guidance on using `Activity` with `DiagnosticSource`:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#starting-and-stopping-activity

In this scenario, a telemetry consumer who chooses to listen to `"Http_In"`-evens will receive _Start_ and _Stop_ events such that the _Start_ events already contain information from the `"Correlation-Context"`-headers in the `Baggage` bag attached to the activity. (In addition, the whole `HttpContext context` object is also passed, but this is not the focus here.)

On contrary, in the new `ActivitySource`-based APIs, the listeners are notified within the scope of the `ActivitySource.StartActivity(..)` method. The creator of the activity only receives a reference to the new activity after the potentially registered listeners have been notified and finished processing the _Start_ event. Only information that can be passed as arguments to the `StartActivity(..)` method can reach the listeners.

So the scenario that is used as the "model" scenario in .NET's own docs is not supported.

_If we consider this to indeed be a problem worth addressing, we should not stop at `Baggage` and consider that it applies to <ins>all</ins> _Activity_ properties that existed pre-.NET-5 and were settable on an `Activity` instance before is is passed to `DiagnosticSource.StartActivity(..)`._

I realized how late to the final release the product is. There are several possible approaches:

* I am missing something, and there is a workaround already (I hope for this to be the case :) ). Perhaps we can include it into some doc.

* We decide that this is indeed a problem, but we are too late in the cycle to fix it. The we should still acknowledge and document it, and track for a future release.

* We consider it fixing for .NET 5. If so, please let me know, I'd be happy to prepare a PR.

Thanks for considering this!

-----------

**(Update an hour later)** Just discovered an additional (related) issue, adding relevant info here:

For convenience, here is the current [ActivitySource public surface](https://github.com/dotnet/runtime/blob/92180ed3e98bd6153124d6a80095cefa31267711/src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs#L131).

The `Tags` of an `Activity` are not exactly affected by the above problem as `ActivitySource.StartActivity(..)` has overloads that take an `IEnumerable<KeyValuePair<string, object>> tags` parameter. However, this will require that an enumerable is allocated before `StartActivity(..)` is called. This is problematic in two ways. One of them is easier to address than the other:

1. If the activity would _not_ be created due to sampling (i.e. `StartActivity(..)` returns `null`), then the allocation of the enumerable and _all_ its members was for nothing. This is _hard_ to address in the current model. At this stage all we can do is to document it and recommend a best way to minimize overhead.

2. Even if the Activity object _is_ created, there is still an allocation if the _IEnumerable_ just to pass parameters. The object will likely not survive beyond Gen0, but still: so much work was done to avoid allocations, that it seems weird. This could be addressed by providing `StartActivity(..)` overloads that have parameters like `string tag1Key`, `string tag1Value`, ... for the first few tags. This could help skipping that allocation for the common case where there is just a few tags. Notably, having tags and/or baggage is likely much more frequent than having links, events etc. So yes, the number of overloads would grow, but would not explode uncontrollably. :)

**Comment by terrajobst** [comment]
* We should add a simplified overload that only requires name and kind

```C#
namespace System.Diagnostics
{
    public sealed class ActivitySource : IDisposable
    {
        public Activity? CreateActivity(string name, 
                                        ActivityKind kind);

        public Activity? CreateActivity(string name,
                                        ActivityKind kind, 
                                        ActivityContext parentContext,
                                        IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                                        IEnumerable<ActivityLink>? links = null);

        public Activity? CreateActivity(string name, 
                                        ActivityKind kind,
                                        string parentContext,
                                        IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                                        IEnumerable<ActivityLink>? links = null);
    }
}
```

**Comment by tarekgh** [comment]
Discussed offline more the proposal with @dotnet/fxdc to ensure we are addressing the request in https://github.com/dotnet/runtime/issues/43853 and decided we'll add the extra parameter `ActivityIdFormat idFormat = ActivityIdFormat.Unknown` to the last 2 overloads. the final approved proposal now is: 

```C#
namespace System.Diagnostics
{
    public sealed class ActivitySource : IDisposable
    {
        public Activity? CreateActivity(string name, 
                                        ActivityKind kind);

        public Activity? CreateActivity(string name,
                                        ActivityKind kind, 
                                        ActivityContext parentContext,
                                        IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                                        IEnumerable<ActivityLink>? links = null,
                                        ActivityIdFormat idFormat = ActivityIdFormat.Unknown);

        public Activity? CreateActivity(string name, 
                                        ActivityKind kind,
                                        string parentContext,
                                        IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                                        IEnumerable<ActivityLink>? links = null,
                                        ActivityIdFormat idFormat = ActivityIdFormat.Unknown);
    }
}
```
